### PR TITLE
BUG: Fix unwanted type casting while replacing values in a DataFrame

### LINKED ISSUE #32988: REGR: `replace` casts columns to `object`
#### Calling `df.replace` casts columns to `object`

```python
pd.DataFrame(np.eye(2)).replace(to_replace=[None, -np.inf, np.inf], value=pd.NA).dtypes
# 0    object
# 1     object
# dtype: object
```
#### Problem description
I'd expect that the dtypes of the columns remain the same after replacing with pd.NA (especially considering no values are replaced in the above call). We do not get this issue if `to_replace` is any subset of `[None, -np.inf, np.inf]`. We get the same issue if `value` is instead `np.nan`.

#### Expected Output
```python
# 0    float64
# 1     float64
# dtype: object
```
#### Output of ``pd.show_versions()``

<details>
Installed with `pip install -e .` on current master. Was also able to reproduce on `1.0.1`.

INSTALLED VERSIONS
------------------
commit           : bed9103e52ae412774c22b8e198f7b66708981c7
python           : 3.7.0.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
Version          : Darwin Kernel Version 18.7.0: Tue Aug 20 16:57:14 PDT 2019; root:xnu-4903.271.2~2/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.26.0.dev0+2701.gbed9103e5
numpy            : 1.18.2
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.1.1.post20200323
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.13.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #32988
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

As mentioned in #32988 [here](https://github.com/pandas-dev/pandas/issues/32988#issuecomment-604727057) I think, that I have found a way to fix this, but I don't know, if this results in a desired behavior. 

I would add tests after ensuring, that this does not break anything else. 

Any thoughts about this?

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
always tests first.

**Comment by phofl** [comment]
Thank, I'll follow this way in the future. 

I added a test.

When you use a value, which is actually replaced in the test, the pd.NA is converted to np.nan, if the original DataFrame had float64 as datatype. The example given in the issue did not replace a value from the original DataFrame.

Don't know, if this is a desired behavior.

File: pandas/tests/frame/methods/test_replace.py [hunk 1]
```diff
@@ -22,6 +22,25 @@ def mix_abc() -> Dict[str, List[Union[float, str]]]:
 
 
 class TestDataFrameReplace:
+    def test_replace_without_type_cast(self):
+        # PH: 32988: Fix unwanted type casting while replacing
+        result = (
+            pd.DataFrame(np.eye(2))
+            .replace(to_replace=[None, -np.inf, np.inf], value=pd.NA)
+            .dtypes
```
Comment (jreback) [inline pandas/tests/frame/methods/test_replace.py]: don't check dtypes but actual values

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -734,14 +734,25 @@ def replace(
...
+            block_replaced = block.replace(
                 to_replace=to_replace,
                 value=value,
                 inplace=inplace,
                 filter=filter,
                 regex=regex,
                 convert=convert,
             )
+            blocks_converted = []
+            for ls_elem in block_replaced:
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: why is this logic not just in Block.replace itself? seems too complicated here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/methods/test_replace.py (hunk 1 above)
Comment (phofl) [reply pandas/tests/frame/methods/test_replace.py]: Changed the test

File: pandas/core/internals/blocks.py (hunk 2 above)
Comment (phofl) [reply pandas/core/internals/blocks.py]: I wanted to ensure, that the conversion would only take place, after the block was forcefully cast to an object before.

I can add it to the replace function, if that would be better.

**Comment by simonjayhawkins** [comment]
i've labelled this as a regression, see https://github.com/pandas-dev/pandas/issues/32988#issuecomment-605519281 cc @jbrockmendel

**Comment by jreback** [comment]
can you merge master and see if you can simplify

File: pandas/core/internals/blocks.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/internals/blocks.py:722]: > I wanted to ensure, that the conversion would only take place, after the block was forcefully cast to an ObjectBlock before.
> 
> I can add it to the replace function, if that would be better.

the regression was caused by a clean-up to Block.replace (#27768), see https://github.com/pandas-dev/pandas/issues/32988#issuecomment-605519281

so fixing Block.replace as @jreback suggests seems more appropriate

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]

**Comment by simonjayhawkins** [comment]
Thanks @phofl for looking into this. #32988 has been addressed via #34048.

**Pull request closed without merging**